dali/python/nvidia/dali/plugin/jax.py

```diff
@@ -181,45 +191,62 @@ def __next__(self):
             self._first_batch = None
             return batch
 
-        # Gather outputs
         pipelines_outputs = self._get_outputs()  # Can be accessed by outputs[device_id][output_id]
 
         next_output = dict()
         for category_id, category_name in enumerate(self.output_map):
-            category_outputs = []
-
-            # Gather outputs for current category from all pipelines
-            for pipeline_id in range(self._num_gpus):
-                category_outputs.append(
-                    _to_jax_array(pipelines_outputs[pipeline_id][category_id].as_tensor()))
+            category_outputs = self._gather_outputs_for_category(
+                pipelines_outputs, category_id)
 
             if self._num_gpus == 1:
                 next_output[category_name] = category_outputs[0]
-            else:   # Assemble output from multiple pipelines
-                for shard in category_outputs:
-                    assert shard.shape == category_outputs[0].shape, \
-                        "Shards shapes have to be the same."
-
-                category_outputs_devices = tuple(map(
-                    lambda jax_shard: jax_shard.device(),
-                    category_outputs))
-
-                distinct_category_outputs_devices = set(category_outputs_devices)
-
-                if len(category_outputs_devices) != len(distinct_category_outputs_devices):
-                    if len(distinct_category_outputs_devices) != 1:
-                        raise AssertionError("JAX iterator requires shards to be placed on \
-                                             different devices or all on the same device.")
-                    else:
-                        # All shards are on one device.
-                        next_output[category_name] = jnp.stack(category_outputs)
+            else:
+                self._assert_shards_shapes(category_outputs)
+                if self._sharding is not None:
+                    next_output[category_name] = self._build_output_with_sharding(category_outputs)
                 else:
-                    # Build sharded JAX array as output for current category
-                    next_output[category_name] = jax.device_put_sharded(
-                        category_outputs,
-                        category_outputs_devices)
+                    next_output[category_name] = self._build_output_with_devices(
+                        next_output, category_name, category_outputs)
 
         self._schedule_runs()
         self._advance_and_check_drop_last()
 
         return next_output
+
+    def _gather_outputs_for_category(self, pipelines_outputs, category_id):
+        category_outputs = []
+
+        for pipeline_id in range(self._num_gpus):
+            category_outputs.append(
+                _to_jax_array(pipelines_outputs[pipeline_id][category_id].as_tensor()))
+
+        return category_outputs
+
+    def _build_output_with_devices(self, next_output, category_name, category_outputs):
```

Comment:
I don't quite follow what this function does. An error message in L234 mentions sharding, but the function is invoked when `_sharding is None`.
Some comment explaining high level functionality would be nice to avoid making this code "write only".

Reply:
Generally, there are two mechanisms of building sharded output: `jax.sharding.Sharding` and `jax.device_put_sharded()`. We want to support both. The error mentioned `shard` as a general concept not `sharding` object.
 
I changed the name to `_build_output_with_device_put` to indicate which function takes which path.